packages/tpetra/core/ext/TpetraExt_MatrixMatrix_decl.hpp

```diff
@@ -648,6 +542,105 @@ void setMaxNumEntriesPerRow(
+  static void addSorted(
+    const values_array& Avals,
+    const row_ptrs_array_const& Arowptrs,
+    const col_inds_array& Acolinds,
+    const impl_scalar_type scalarA,
+    const values_array& Bvals,
+    const row_ptrs_array_const& Browptrs,
+    const col_inds_array& Bcolinds,
+    const impl_scalar_type scalarB,
+    values_array& Cvals,
+    row_ptrs_array& Crowptrs,
+    col_inds_array& Ccolinds);
+
+  /// \brief Given two matrices in CRS format, return their sum
+  /// \pre A and B don't need to be sorted, and column indices are given as global indices
+  /// \param A A (local) matrix
+  /// \param scalarA Scaling factor for A
+  /// \param B B (local) matrix
+  /// \param scalarB Scaling factor for B
+  /// \param minGlobalCol The minimum global index owned by this processor in the column map
+  /// \param globalNumCols The global size of the column map
+  /// \param[Out] Cvals Values array for C (allocated inside function)
+  /// \param[Out] Crowptrs Row pointers array for C (allocated inside function)
+  /// \param[Out] Ccolinds Column indices array for C (allocated inside function)
+  static void convertToGlobalAndAdd(
+    const KCRS& A,
+    const impl_scalar_type scalarA,
+    const KCRS& B,
+    const impl_scalar_type scalarB,
+    const local_map_type& AcolMap,
+    const local_map_type& BcolMap,
+    GlobalOrdinal minGlobalCol,
+    GlobalOrdinal globalNumCols,
+    values_array& Cvals,
+    row_ptrs_array& Crowptrs,
+    global_col_inds_array& Ccolinds);
+
+  /// \brief Given two matrices in CRS format, return their sum
+  /// \pre A and B don't need to be sorted
+  /// \param Avals Values array for A
+  /// \param Arowptrs Row pointers array for A
+  /// \param Acolinds Column indices array for A
+  /// \param scalarA Scaling factor for A
+  /// \param Bvals Values array for B
+  /// \param Browptrs Row pointers array for B
+  /// \param Bcolinds Column indices array for B
+  /// \param scalarB Scaling factor for B
+  /// \param globalNumCols The global size of the column map
+  /// \param[Out] Cvals Values array for C (allocated inside function)
+  /// \param[Out] Crowptrs Row pointers array for C (allocated inside function)
+  /// \param[Out] Ccolinds Column indices array for C (allocated inside function)
+  static void addUnsorted(
+    const values_array& Avals,
+    const row_ptrs_array_const& Arowptrs,
+    const col_inds_array& Acolinds,
+    const impl_scalar_type scalarA,
+    const values_array& Bvals,
+    const row_ptrs_array_const& Browptrs,
+    const col_inds_array& Bcolinds,
+    const impl_scalar_type scalarB,
+    GlobalOrdinal numGlobalCols,
+    values_array& Cvals,
+    row_ptrs_array& Crowptrs,
+    col_inds_array& Ccolinds);
+
+  static Teuchos::RCP<map_type> makeColMapAndConvertGids(GlobalOrdinal indexBase, GlobalOrdinal minCol, GlobalOrdinal maxCol, const global_col_inds_array& gids, col_inds_array& lids, const Teuchos::RCP<const Teuchos::Comm<int>>& comm);
```

Comment:
`Tpetra::Map` is immutable, so if you return it by pointer, the pointer should always be a pointer-to-const (e.g., `RCP<const map_type>`).